File: examples/nodejs/azure-secrets.js
/**
 * Azure Key Vault Integration
 *
 * WHY AZURE KEY VAULT OVER .env FILES:
 * ────────────────────────────────────
 * Azure Key Vault provides centralized secret management with:
 *   - Hardware Security Modules (HSMs) for FIPS 140-2 Level 2 / Level 3 protection
 *   - Automatic certificate management and renewal
 *   - Role-Based Access Control (RBAC) via Azure AD / Entra ID
 *   - Soft-delete and purge protection (accidental deletion recovery)
 *   - Network isolation via Private Endpoints (no public internet exposure)
 *
 * AUTHENTICATION: We use DefaultAzureCredential which tries multiple auth
 * methods in order: Environment variables → Managed Identity → VS Code →
 * Azure CLI → Azure PowerShell. This means the same code works locally
 * (with az login) and in production (with Managed Identity) with zero changes.
 */

const { SecretClient } = require("@azure/keyvault-secrets");
const { DefaultAzureCredential } = require("@azure/identity");
const winston = require("winston");

// ── Logger Configuration ──────────────────────────────────────────────────
// SECURITY RULE: Never log secret values. Always use [REDACTED].
const logger = winston.createLogger({
  level: process.env.LOG_LEVEL || "info",
  format: winston.format.combine(
    winston.format.timestamp(),
    winston.format.errors({ stack: true }),
    winston.format.json()
  ),
  transports: [new winston.transports.Console()],
});

// ── Azure Key Vault Client Initialization ─────────────────────────────────
let secretClient;

/**
 * Creates and returns an Azure Key Vault SecretClient.
 *
 * BOOTSTRAPPING FROM ENVIRONMENT:
 *   AZURE_VAULT_URL - The Key Vault URI (e.g., https://myapp-vault.vault.azure.net/)
 *
 * AUTHENTICATION CHAIN (DefaultAzureCredential tries in order):
 *   1. AZURE_TENANT_ID, AZURE_CLIENT_ID, AZURE_CLIENT_SECRET env vars
 *   2. Azure Managed Identity (for VMs, App Service, AKS, etc.)
 *   3. Visual Studio Code Azure authentication
 *   4. Azure CLI (az login) — useful for local development
 *   5. Azure PowerShell — useful for local development
 *
 * This chain means developers can use `az login` locally while production
 * uses Managed Identity — no code changes required.
 */
function getClient() {
  if (!secretClient) {
    const vaultUrl = process.env.AZURE_VAULT_URL;

    if (!vaultUrl) {
      throw new Error(
        "AZURE_VAULT_URL environment variable is required. " +
        "Example: https://myapp-vault.vault.azure.net/"
      );
    }

    // Validate the URL format to catch misconfigurations early
    if (!vaultUrl.startsWith("https://")) {
      logger.warn("AZURE_VAULT_URL should use HTTPS", { vaultUrl });
    }

    // DefaultAzureCredential automatically selects the right auth method
    // based on the runtime environment. No hardcoded credentials needed.
    const credential = new DefaultAzureCredential();

    secretClient = new SecretClient(vaultUrl, credential);

    logger.info("Azure Key Vault client initialized", {
      vaultUrl,
      // We don't log which credential type was selected to avoid leaking
      // infrastructure details, but in debug mode this is useful
    });

    // Verify connectivity by checking if the vault is accessible
    // (This is a lightweight operation — it doesn't fetch any secrets)
    credential
      .getToken("https://vault.azure.net/.default")
      .then(() => logger.debug("Azure credential successfully acquired"))
      .catch((err) => {
        logger.error("Failed to acquire Azure credential — check your auth setup", {
          error: err.message,
        });
      });
  }
  return secretClient;
}

/**
 * Fetches a secret by name from Azure Key Vault.
 *
 * Azure Key Vault stores each secret as a named value with versioning.
 * By default, getSecret() returns the latest version. You can pin to
 * a specific version by appending "?version=xyz" — but latest is
 * recommended so you pick up rotations automatically.
 *
 * @param {string} secretName - The name of the secret in Key Vault.
 * @param {string} [version] - Optional specific version ID.
 * @returns {Promise<string>} The secret value.
 */
async function getSecret(secretName, version) {
  const client = getClient();

  logger.info("Fetching secret from Azure Key Vault", {
    secretName,
    version: version || "latest",
  });

  try {
    const options = version ? { version } : {};
    const response = await client.getSecret(secretName, options);

    logger.info("Secret retrieved successfully", {
      secretName,
      version: response.properties?.version || "unknown",
      value: "[REDACTED]",
      enabled: response.properties?.enabled,
      expiresOn: response.properties?.expiresOn || "never",
    });

    return response.value;
  } catch (error) {
    if (error.statusCode === 404) {
      logger.error("Secret not found in Azure Key Vault", {
        secretName,
        version: version || "latest",
      });
      throw new Error(`Secret '${secretName}' not found in Key Vault`);
    }
    if (error.statusCode === 403) {
      logger.error("Access denied to Azure Key Vault secret — check RBAC / Access Policies", {
        secretName,
        error: error.message,
      });
      throw new Error(`Access denied for secret '${secretName}'`);
    }
    logger.error("Failed to fetch secret from Azure Key Vault", {
      secretName,
      error: error.message,
    });
    throw error;
  }
}

/**
 * Fetches database connection string from Key Vault.
 *
 * Azure convention: Store connection strings as a single secret value
 * rather than separate username/password/host fields. This matches how
 * Azure services (App Service, Functions) inject connection strings.
 *
 * ROTATION SUPPORT: When you update a secret in Key Vault (new version),
 * the application fetches the new value on the next getSecret() call.
 * For zero-downtime rotation, implement a secret refresh interval.
 *
 * @param {string} secretName - Secret name (default: "DatabaseConnectionString")
 * @returns {Promise<string>} The connection string.
 */
async function getDatabaseConnectionString(secretName = "DatabaseConnectionString") {
  const connectionString = await getSecret(secretName);

  if (!connectionString) {
    logger.error("Database connection string is empty", { secretName });
    throw new Error(`Secret '${secretName}' returned an empty connection string`);
  }

  // SECURITY: Log that we got the string, but never the string itself.
  // Even connection strings can contain embedded credentials.
  logger.info("Database connection string retrieved", {
    secretName,
    length: connectionString.length,
    hasPassword: connectionString.includes("Password=") ||
                 connectionString.includes("pwd=") ||
                 connectionString.includes("password="),
    prefix: connectionString.substring(0, connectionString.indexOf("://") + 3),
  });

  return connectionString;
}

/**
 * Fetches JWT signing key from Key Vault.
 *
 * @param {string} secretName - Secret name (default: "JwtSigningKey")
 * @returns {Promise<string>} The JWT secret key.
 */
async function getJwtSigningKey(secretName = "JwtSigningKey") {
  const key = await getSecret(secretName);

  if (!key) {
    logger.error("JWT signing key is empty", { secretName });
    throw new Error(`Secret '${secretName}' returned an empty JWT key`);
  }

  logger.info("JWT signing key retrieved", {
    secretName,
    keyLength: key.length,
    value: "[REDACTED]",
  });

  return key;
}

/**
 * Fetches an API key from Key Vault.
 *
 * Many third-party services require API keys. Storing them in Key Vault
 * means you can rotate keys without redeploying. Azure Key Vault also
 * supports secret expiration dates — set one to ensure keys are reviewed
 * and rotated on a regular schedule.
 *
 * @param {string} secretName - Secret name for the API key.
 * @returns {Promise<string>} The API key value.
 */
async function getApiKey(secretName) {
  const apiKey = await getSecret(secretName);

  if (!apiKey) {
    logger.error("API key is empty", { secretName });
    throw new Error(`Secret '${secretName}' returned an empty API key`);
  }

  // Log a prefix (e.g., "sk-proj...") to aid debugging without exposing
  // the full key. Most API keys have recognizable prefixes.
  const prefix = apiKey.length > 8 ? `${apiKey.substring(0, 7)}...` : "[REDACTED]";

  logger.info("API key retrieved", {
    secretName,
    apiKeyPrefix: prefix,
    keyLength: apiKey.length,
  });

  return apiKey;
}

module.exports = {
  getSecret,
  getDatabaseConnectionString,
  getJwtSigningKey,
  getApiKey,
  getClient,
};

File: examples/python/azure_secrets.py
"""
Azure Key Vault Integration Module
====================================
This module demonstrates how to securely fetch secrets from Azure Key Vault
using the azure-identity and azure-keyvault-secrets SDKs.

Security rationale
------------------
* Authentication uses `DefaultAzureCredential`, which tries multiple identity
  providers in order (Managed Identity → Environment Variables → Visual
  Studio Code → Azure CLI → Interactive Browser).  This means:
  - On Azure (VMs, App Service, AKS): Managed Identity is used automatically
    with NO credentials in the environment at all — the identity is bound to
    the compute resource.
  - In local dev: Azure CLI or VS Code credentials are picked up.
  - No service principals or client secrets need to be stored anywhere.
* The Key Vault URL is the only configuration needed, sourced from the
  AZURE_VAULT_URL environment variable.
* Key Vault supports secret versioning, expiration dates, and access
  policies that restrict which identities can read which secrets.
* All secret values remain in memory and are never logged.

Prerequisites
-------------
  pip install azure-identity azure-keyvault-secrets
"""

import json
import logging
import os
from datetime import datetime, timezone
from typing import Any

from azure.core.exceptions import HttpResponseError, ResourceNotFoundError
from azure.identity import DefaultAzureCredential
from azure.keyvault.secrets import KeyVaultSecret, SecretClient

# ---------------------------------------------------------------------------
# Structured logging
# ---------------------------------------------------------------------------
logger = logging.getLogger("azure_secrets")
logger.setLevel(logging.INFO)


def _log_event(event: str, secret_name: str | None = None, **extra: Any) -> None:
    """Emit a structured log entry, redacting any sensitive keys."""
    payload = {"event": event, "secret_name": secret_name, **extra}
    for k, v in payload.items():
        if any(s in k.lower() for s in ("key", "token", "secret", "password", "value")):
            payload[k] = "***REDACTED***"
    logger.info(payload)


class AzureKeyVaultManager:
    """High-level wrapper around Azure Key Vault for secret retrieval.

    Each secret in Key Vault can be a simple string or a JSON blob.
    For structured config (DB creds, JWT settings, OpenAI keys) we store
    a JSON string and parse it here, giving us typed access.
    """

    DEFAULT_OPENAI_SECRET = "openai-config"
    DEFAULT_DB_SECRET = "database-credentials"
    DEFAULT_JWT_SECRET = "jwt-config"

    def __init__(
        self,
        vault_url: str | None = None,
        openai_secret_name: str | None = None,
        db_secret_name: str | None = None,
        jwt_secret_name: str | None = None,
    ) -> None:
        # ------------------------------------------------------------------
        # Bootstrapping from environment variables:
        #   AZURE_VAULT_URL    — e.g. "https://my-vault.vault.azure.net/"
        #   AZURE_TENANT_ID    — optional, helps DefaultAzureCredential
        #   AZURE_CLIENT_ID    — optional, for service principal fallback
        #   AZURE_CLIENT_SECRET — optional, for service principal fallback
        #
        # SECURITY: Prefer Managed Identity when running on Azure.
        # Service principal credentials are a fallback for environments
        # where Managed Identity is not available.
        # ------------------------------------------------------------------
        self._vault_url = vault_url or os.environ["AZURE_VAULT_URL"]

        _log_event(
            "azure_kv_init",
            extra={
                "vault_url": self._vault_url,
                "tenant_id_set": bool(os.getenv("AZURE_TENANT_ID")),
                "client_id_set": bool(os.getenv("AZURE_CLIENT_ID")),
                "client_secret_set": bool(os.getenv("AZURE_CLIENT_SECRET")),
            },
        )

        # DefaultAzureCredential automatically selects the right auth
        # method based on the runtime environment.
        self._credential = DefaultAzureCredential()

        # Verify connectivity eagerly
        try:
            self._client = SecretClient(
                vault_url=self._vault_url,
                credential=self._credential,
            )
            # The Azure SDK is lazy — force a lightweight call to confirm
            # auth works (get_properties_of_secrets is a list operation).
            list(self._client.get_properties_of_secrets().by_page())
        except HttpResponseError as exc:
            _log_event("azure_kv_auth_failed", error=str(exc))
            raise

        # Resolve secret names
        self._openai_secret = (
            openai_secret_name
            or os.getenv("AZURE_SECRET_OPENAI", self.DEFAULT_OPENAI_SECRET)
        )
        self._db_secret = (
            db_secret_name
            or os.getenv("AZURE_SECRET_DB", self.DEFAULT_DB_SECRET)
        )
        self._jwt_secret = (
            jwt_secret_name
            or os.getenv("AZURE_SECRET_JWT", self.DEFAULT_JWT_SECRET)
        )

    # ------------------------------------------------------------------
    # Core fetch
    # ------------------------------------------------------------------
    def get_secret(self, secret_name: str, version: str | None = None) -> str:
        """Fetch a raw secret value by name.

        Args:
            secret_name: The name of the secret in Key Vault.
            version:     Optional version ID (a GUID).  Omit for latest.

        Returns:
            The raw string value of the secret.

        Raises:
            ResourceNotFoundError: secret does not exist.
            HttpResponseError:     permission denied or service error.
        """
        _log_event("fetching_secret", secret_name=secret_name, version=version)

        try:
            if version:
                secret: KeyVaultSecret = self._client.get_secret(
                    secret_name, version=version
                )
            else:
                secret = self._client.get_secret(secret_name)
        except ResourceNotFoundError:
            _log_event("secret_not_found", secret_name=secret_name)
            raise
        except HttpResponseError as exc:
            _log_event(
                "fetch_secret_error",
                secret_name=secret_name,
                status_code=exc.status_code,
            )
            raise

        _log_event(
            "secret_fetched",
            secret_name=secret_name,
            version=secret.properties.version,
            expires_on=str(secret.properties.expires_on or "never"),
        )
        return secret.value

    def get_secret_json(self, secret_name: str, version: str | None = None) -> dict[str, Any]:
        """Fetch a secret and parse it as JSON.

        Key Vault stores everything as strings.  For structured configuration
        we store JSON and parse it here.  This gives us type-safe access.
        """
        raw = self.get_secret(secret_name, version)
        try:
            return json.loads(raw)
        except json.JSONDecodeError as exc:
            raise ValueError(
                f"Secret '{secret_name}' is not valid JSON"
            ) from exc

    # ------------------------------------------------------------------
    # Typed convenience methods
    # ------------------------------------------------------------------
    def get_database_credentials(self) -> dict[str, Any]:
        """Return database connection parameters from Key Vault.

        Expected JSON shape:
        {
            "engine": "postgresql",
            "host": "mydb.postgres.database.azure.com",
            "port": 5432,
            "username": "app_user",
            "password": "<sensitive>",
            "dbname": "app_db",
            "ssl_mode": "require"
        }
        """
        creds = self.get_secret_json(self._db_secret)
        required = {"host", "username", "password", "dbname"}
        missing = required - set(creds.keys())
        if missing:
            raise ValueError(f"DB secret missing fields: {missing}")
        return creds

    def get_jwt_config(self) -> dict[str, Any]:
        """Return JWT signing configuration from Key Vault.

        Expected JSON shape:
        {
            "algorithm": "RS256",
            "secret_key": "<signing-key>",
            "access_token_ttl_minutes": 15,
            "refresh_token_ttl_days": 7
        }
        """
        config = self.get_secret_json(self._jwt_secret)
        if "secret_key" not in config:
            raise ValueError("JWT secret missing 'secret_key' field")
        return config

    def get_openai_config(self) -> dict[str, Any]:
        """Return OpenAI API configuration from Key Vault.

        Expected JSON shape:
        {
            "api_key": "sk-...",
            "organization_id": "org-...",
            "default_model": "gpt-4"
        }
        """
        config = self.get_secret_json(self._openai_secret)
        if "api_key" not in config:
            raise ValueError("OpenAI secret missing 'api_key' field")
        return config

    # ------------------------------------------------------------------
    # Metadata helpers
    # ------------------------------------------------------------------
    def get_secret_metadata(self, secret_name: str) -> dict[str, Any]:
        """Return metadata about a secret (not the value itself).

        Useful for auditing: when was it created/updated?  Does it expire?
        """
        props = self._client.get_secret_properties(secret_name)
        return {
            "name": secret_name,
            "version": props.version,
            "created_on": str(props.created_on),
            "updated_on": str(props.updated_on),
            "expires_on": str(props.expires_on or "never"),
            "enabled": props.enabled,
            "content_type": props.content_type,
        }


# ---------------------------------------------------------------------------
# Quick-start demonstration
# ---------------------------------------------------------------------------
if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO, format="%(message)s")

    try:
        mgr = AzureKeyVaultManager()
        db = mgr.get_database_credentials()
        print(f"DB host: {db['host']}")
        meta = mgr.get_secret_metadata(mgr._db_secret)
        print(f"DB secret version: {meta['version']}")
    except KeyError as exc:
        print(f"AZURE_VAULT_URL not set (expected in local dev): {exc}")
    except Exception as exc:
        print(f"Azure Key Vault error (expected in local dev): {exc}")